numba_rvsdg/core/datastructures/ast_transforms.py

```diff
@@ -0,0 +1,603 @@
+                for b in list(self.values()):
+                    if len(b.jump_targets) == 0:
+                        continue
+                    elif len(b.jump_targets) == 1:
+                        if b.jump_targets[0] == name:
+                            b.jump_targets[0] = it
+                    elif len(b.jump_targets) == 2:
+                        if b.jump_targets[0] == name:
+                            b.jump_targets[0] = it
+                        elif b.jump_targets[1] == name:
+                            b.jump_targets[1] = it
+        self.empty = empty
+        return empty
+
+
+@dataclass(frozen=True)
+class LoopIndices:
+    """Structure to hold the head and exit block indices of a loop."""
+
+    head: int
+    exit: int
+
+
+class AST2SCFGTransformer:
+    """AST2SCFGTransformer
+
+    The AST2SCFGTransformer class is responsible for transforming code in the
+    form of a Python Abstract Syntax Tree (AST) into CFG/SCFG.
+
+    """
+
+    def __init__(self, code: Callable[..., Any], prune: bool = True) -> None:
+        # Prune empty and unreachable blocks from the CFG.
+        self.prune: bool = prune
+        # Save the code for transformation.
+        self.code: Callable[..., Any] = code
+        # Monotonically increasing block index, 0 is reserved for genesis.
+        self.block_index: int = 1
+        # Dict mapping block indices as strings to WritableASTBlocks.
+        # (This is the data structure to hold the CFG.)
+        self.blocks: ASTCFG = ASTCFG()
+        # Initialize first (genesis) block, assume it's named zero.
+        # (This also initializes the self.current_block attribute.)
+        self.add_block(0)
+        # Stack for header and exiting block of current loop.
+        self.loop_stack: list[LoopIndices] = []
+
+    def transform_to_ASTCFG(self) -> ASTCFG:
+        """Generate ASTCFG from Python function."""
+        self.transform()
+        return self.blocks
+
+    def transform_to_SCFG(self) -> SCFG:
+        """Generate SCFG from Python function."""
+        self.transform()
+        return self.blocks.to_SCFG()
+
+    def add_block(self, index: int) -> None:
+        """Create block, add to CFG and set as current_block."""
+        self.blocks[str(index)] = self.current_block = WritableASTBlock(
+            name=str(index)
+        )
+
+    def seal_block(self, default_index: int) -> None:
+        """Seal the current block by setting the jump_targets."""
+        if self.loop_stack:
+            self.current_block.seal_inside_loop(
+                self.loop_stack[-1].head,
+                self.loop_stack[-1].exit,
+                default_index,
+            )
+        else:
+            self.current_block.seal_outside_loop(default_index)
+
+    def transform(self) -> None:
+        """Transform Python function stored as self.code."""
+        # Convert source code into AST.
+        tree = ast.parse(textwrap.dedent(inspect.getsource(self.code))).body
```

Comment:
The expected input should be an AST instead of code. Otherwise, if I start with a source string, I'll need to make a code object first. I'll suggest pull out this AST creation logic.

Reply:
For example, i tried to use my [AST fuzzer](https://github.com/sklam/jittery/blob/pythrankenstein/pythrankenstein.py) to test this but i ended up needing to change the code like:

```diff
@@ -221,7 +221,10 @@ class AST2SCFGTransformer:
     def transform(self) -> None:
         """Transform Python function stored as self.code."""
         # Convert source code into AST.
-        tree = ast.parse(textwrap.dedent(inspect.getsource(self.code))).body
+        if isinstance(self.code, str):
+            tree = ast.parse(self.code).body
+        else:
+            tree = ast.parse(textwrap.dedent(inspect.getsource(self.code))).body
         # Assert that the code handed in was a function, we can only transform
         # functions.
         assert isinstance(tree[0], ast.FunctionDef)
```

Btw, the fuzzer will need `ast.Pass` handled to work correctly.

Reply:
`ast.Pass` can be handled and pruned with the `prune_noop` pass: dc64724bc05890004065a97161d8130a88d7135a

Reply:
I opted to include this snippet such that you can now feed both a callable and a string to the transformer: 17525f17ff64f36dc80f31a78194ef6921bacc3c